### PR TITLE
[RISCV] Add support for inline asm constraint vd

### PR DESCRIPTION
**Pull request by tclin914**
It constrains vector registers excluding v0. Refer to https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html RISC-V part.

This patch also adds a testcase for constraints vr, vd and vm.

### CODE REVIEW

**Comment by lenary** [comment]
I think you've missed making the relevant clang changes to `RISCVTargetInfo::validateAsmConstraint` to support `vd`.

I don't understand why this constraint is needed but it's good to stay in line with GCC.

**Review by lenary — APPROVED** [review]
LGTM

File: clang/lib/Basic/Targets/RISCV.cpp [hunk 1]
```diff
@@ -102,7 +102,7 @@ bool RISCVTargetInfo::validateAsmConstraint(
     return true;
   case 'v':
     // A vector register.
-    if (Name[1] == 'r' || Name[1] == 'm') {
+    if (Name[1] == 'r' || Name[1] == 'd' || Name[1] == 'm') {
```
Comment (topperc) [inline clang/lib/Basic/Targets/RISCV.cpp:105]: This needs to be tested in clang/test/CodeGen/RISCV/riscv-inline-asm-rvv.c

**Review by topperc — CHANGES_REQUESTED** [review]

File: clang/test/CodeGen/RISCV/riscv-inline-asm-rvv.c [hunk 2]
```diff
@@ -29,6 +29,14 @@ vint32m1_t test_vr(vint32m1_t a, vint32m1_t b) {
   return ret;
 }
 
+vint32m1_t test_vd(vint32m1_t a, vint32m1_t b) {
+// CHECK-LABEL: define{{.*}} @test_vd
+// CHECK: %0 = tail call <vscale x 2 x i32> asm sideeffect "vadd.vv $0, $1, $2", "=^vd,^vd,^vd"(<vscale x 2 x i32> %a, <vscale x 2 x i32> %b)
+  vint32m1_t ret;
+  asm volatile ("vadd.vv %0, %1, %2" : "=vd"(ret) : "vd"(a), "vd"(b));
```
Comment (lenary) [inline clang/test/CodeGen/RISCV/riscv-inline-asm-rvv.c:36]: @topperc is this test not enough?

Comment (topperc) [reply clang/test/CodeGen/RISCV/riscv-inline-asm-rvv.c:36]: weird how did i not see that was here.

**Review by topperc — APPROVED** [review]
LGTM

File: clang/lib/Basic/Targets/RISCV.cpp (hunk 1 above)
Comment (topperc) [reply clang/lib/Basic/Targets/RISCV.cpp:105]: Nevermind. I guess I'm blind.

**Pull request merged**